### PR TITLE
Horizontal layout

### PR DESCRIPTION
**Pull request by sarcevic-luka**
Added horizontal filling

### CODE REVIEW

File: CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift [hunk 1]
```diff
@@ -10,20 +10,27 @@ import UIKit
...
+}
+
 final class HomeCollectionViewLayout: UICollectionViewLayout {
   let sectionInsets = UIEdgeInsets(top: 16, left: 16, bottom: 16, right: 16)
-  let itemSize = CGSize(width: 100, height: 45)
+  let itemSize = CGSize(width: 120, height: 45)
   let itemSpacing: CGFloat = 8
   private var itemAttributes = Set<UICollectionViewLayoutAttributes>()
   private var sectionHeights = Set<SectionDescriptor>()
+  weak var delegate: HomeCollectionViewLayoutDelegate?
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift]: Properties should be ordered from most accessible `public`/`internal` to least accessible.

File: CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift [hunk 2]
```diff
@@ -36,6 +43,7 @@ final class HomeCollectionViewLayout: UICollectionViewLayout {
 }
 
 extension HomeCollectionViewLayout {
+
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift]: Redundant spacing.

File: CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift [hunk 3]
```diff
@@ -50,21 +58,36 @@ extension HomeCollectionViewLayout {
...
           itemAttributes.insert(attributes)
         }
       }
     }
   }
   
+  private func numberOfColumns(in section: Int) -> Int {
+    guard let collectionView = collectionView else {
+      return 0
+    }
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift]: We could turn this into a one liner:
`guard let collectionView = collectionView else { return 0 }`

Feel free to keep it as-is if you'd prefer to.

File: CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift [hunk 4]
```diff
@@ -50,21 +58,36 @@ extension HomeCollectionViewLayout {
...
+    return delegate?.collectionViewNumberOfColumns(collectionView, layout: self, for: section) ?? 0
+  }
+  
+  private func numberOfRows(in section: Int) -> Int {
+    guard let collectionView = collectionView else {
+      return 0
+    }
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift]: We could turn this into a one liner:
`guard let collectionView = collectionView else { return 0 }`

Feel free to keep it as-is if you'd prefer to.

File: CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift [hunk 5]
```diff
@@ -50,21 +58,36 @@ extension HomeCollectionViewLayout {
...
     guard let collectionView = collectionView else {
       return .zero
     }
-    let totalNumberOfItems = (0..<collectionView.numberOfSections).reduce(0, { $0 + collectionView.numberOfItems(inSection: $1) })
-    let totalHeight = CGFloat(totalNumberOfItems) * itemSize.height + CGFloat(totalNumberOfItems - 1) * itemSpacing + sectionInsets.top + sectionInsets.bottom
-    return CGSize(width: collectionView.frame.width, height: totalHeight)
+    let totalHeightFromSections = sectionHeights.map({$0.height}).reduce(0, +)
+    let highestWidthFromAllSections = sectionHeights.map{$0.width}.max() ?? collectionView.frame.width
+    let maxWidthValue = max(collectionView.frame.width, highestWidthFromAllSections)
+    return CGSize(width: maxWidthValue , height: totalHeightFromSections)
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift]: Redundant whitespace (spacing):
`return CGSize(width: maxWidthValue , height: totalHeightFromSections)`
-> 
`return CGSize(width: maxWidthValue, height: totalHeightFromSections)`

File: CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift [hunk 6]
```diff
@@ -87,14 +110,18 @@ private extension HomeCollectionViewLayout {
 // MARK: - Utility
 private extension HomeCollectionViewLayout {
   func frameForCell(at indexPath: IndexPath) -> CGRect {
-    let previousSections = 0..<indexPath.section
+    let previousSections = 0..<indexPath.sectionIndex
     let allPreviousSectionHeights = sectionHeights
-      .filter { previousSections.contains($0.index) }
+      .filter { previousSections.contains($0.sectionIndex) }
       .reduce(CGFloat(0), { $0 + $1.height }) + CGFloat(previousSections.count) * itemSpacing
+    let itemPostitionInRow = indexPath.columnIndex % numberOfColumns(in: indexPath.sectionIndex)
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift]: Typo in variable name: `itemPostitionInRow`

File: CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift [hunk 7]
```diff
@@ -87,14 +110,18 @@ private extension HomeCollectionViewLayout {
...
+    let currentRow = CGFloat(indexPath.columnIndex / numberOfColumns(in: indexPath.sectionIndex))
+    let xPosition = sectionInsets.left + CGFloat(itemPostitionInRow) * (itemSize.width + itemSpacing)
+    let yPosition = sectionInsets.top + CGFloat(currentRow * (itemSize.height + itemSpacing)) + allPreviousSectionHeights
     
     let origin = CGPoint(
-      x: sectionInsets.left,
-      y: sectionInsets.top + CGFloat(indexPath.item) * itemSize.height + max(0, CGFloat(indexPath.item) * itemSpacing) + allPreviousSectionHeights
+      x: xPosition,
+      y: yPosition
     )
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/App/Scenes/Home/Layout/HomeCollectionViewLayout.swift]: Multiline `CGPoint` is not really needed here:
`let origin = CGPoint(x: xPosition, y: yPosition)`

File: CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+extension IndexPath {
+  var sectionIndex: Int {
+    return self[0]
+  }
+  
+  var columnIndex: Int {
+    return self[1]
+  }
+  var rowIndex: Int {
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift]: Missing newline.

File: CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    return self[2]
+  }
+
+  init(tableSection: Int, column: Int, rowInSection: Int) {
+    self = IndexPath(item: column, section: tableSection)
+    self.append(rowInSection)
+    }
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift]: Incorrect formatting for closing bracket `}`.

File: CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  var columnIndex: Int {
+    return self[1]
+  }
+  var rowIndex: Int {
+    return self[2]
+  }
+
+  init(tableSection: Int, column: Int, rowInSection: Int) {
+    self = IndexPath(item: column, section: tableSection)
+    self.append(rowInSection)
```
Comment (dkulundzic) [inline CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift]: Redundant `self`.

File: CollectionViewLayout/System/Sources/Protocols/DataSourceProtocol.swift [hunk 11]
```diff
@@ -39,6 +39,6 @@ public extension DataSourceProtocol {
 	}
 	
 	func item(at indexPath: IndexPath) -> Section.Item? {
-		return section(at: indexPath.section)?.item(at: indexPath.item)
+		return section(at: indexPath.sectionIndex)?.item(at: indexPath.columnIndex)
```
Comment (dkulundzic) [inline CollectionViewLayout/System/Sources/Protocols/DataSourceProtocol.swift]: Is the **indentation** off a bit here?

File: CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift (hunk 9 above)
Comment (sarcevic-luka) [reply CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift]: I'm missing to see what's wrong.
What should be the correct formatting?

Comment (dkulundzic) [reply CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift]:
```
iinit(tableSection: Int, column: Int, rowInSection: Int) {
    self = IndexPath(item: column, section: tableSection)
    self.append(rowInSection)
    }
```

--> 

```
init(tableSection: Int, column: Int, rowInSection: Int) {
    self = IndexPath(item: column, section: tableSection)
    self.append(rowInSection)
}
```

Could be that you just need reindent, even though Xcode's showing correct indentation.

Comment (sarcevic-luka) [reply CollectionViewLayout/CollectionViewLayout/Common/Extensions/Foundation/IndexPath-CollectionViewLayout.swift]: Control + I = would help me if I tried it at the start. 
Thanks!

**Pull request merged**